torch/csrc/jit/codegen/cuda/tensor_view.cpp

```diff
@@ -314,6 +316,326 @@ TensorView* TensorView::rFactor(const std::vector<int>& axes) {
+
+  // Get inputs for origin expression
+  auto expr_inputs = origin_expr->inputs();
+
+  // Expr* producer_origin =
+  createExprConsumer(origin_expr, this, producer);
+
+  // Expr* producer_uses =
+  new UnaryOp(UnaryOpType::Set, consumer, producer);
+
+  // ComputeAt - This TV -> Other TV
+  // ComputeAt - New TV (Before) -> This TV -> Other TV
+  if (this->hasComputeAt()) {
+    int axis = static_cast<int>(this->getThisComputeAtAxis());
+    producer->computeAt(this->getComputeAtView(), axis);
+  }
+
+  // Before: Other Producers -> This TV
+  // After:  Other Producers -> New TV (Before) -> This TV
+  // Iterate over origin expression inputs
+  for (Val* v : expr_inputs) {
+    if (v->getValType().value() == ValType::TensorView) {
+      TensorView* origin_input = dynamic_cast<TensorView*>(v);
+      if (origin_input->hasComputeAt() &&
+          origin_input->getComputeAtView()->sameAs(this)) {
+        int axis = static_cast<int>(origin_input->getThisComputeAtAxis());
+        origin_input->computeAt(this, axis);
+      }
+    }
+  }
+  return producer;
+}
+
+// Create a TensorView after the original tensor. A common use case is to write
+// results into shared memory or local registers before moving to global memory.
+// Analogous to TVM Cache_Read
+TensorView* TensorView::cache_after() {
+  FusionGuard fg(this->fusion());
+
+  // Get all the uses for this Tensorview
+  std::unordered_set<Expr*> uses_expr = this->fusion()->unordered_uses(this);
+  TORCH_CHECK(
+      uses_expr.size() > 0,
+      "Error adding cache_after ",
+      this,
+      " its #uses is zero and we restrict using cache_after on an output.");
+
+  // Create Consumer Domain
+  // Keep Broadcast Axis (Permanent)
+  auto root_dom = getRootDomain();
+  std::vector<IterDomain*> root_domain;
+  for (size_t i = 0; i < root_dom.size(); ++i) {
+    if (root_dom[i]->isBroadcast()) {
+      root_domain.push_back(new IterDomain(
+          root_dom[i]->start(),
+          root_dom[i]->extent(),
+          root_dom[i]->parallel_method(),
+          false, false, true));
+    } else if (!root_dom[i]->isBroadcast() && !root_dom[i]->isReduction()) {
+      root_domain.push_back(new IterDomain(
+          root_dom[i]->start(),
+          root_dom[i]->extent(),
+          root_dom[i]->parallel_method()));
+    }
+  }
+
+  // This domain will be the producer, so create the consumer
+  TensorView* consumer = new TensorView(
+      new TensorDomain(root_domain), this->getDataType().value());
+
+  // Set domain of producer - No Change
+  TensorView* producer = this;
+
+  // Setup dependency chain, inserting consumer - Cache_Read - after this op.
+  // Before: This Op -> Next Op
+  // After: This Op -> New Cache_Write Op -> Next Op
+
+  Expr* origin_expr = this->fusion()->origin(this);
+  std::deque<Val*> expr_inputs;
+  if (origin_expr != nullptr) {
+    auto inputs = origin_expr->inputs();
+    inputs.assign(expr_inputs.begin(), expr_inputs.end());
+  }
+
```

Comment:
Sorry, having trouble understanding what is going on in the code block above this comment.